io: IllegalStateException in SslTlsSupport with self signed certificate

In 1.1 RC2 we consistently get the following exception with one specific host whilst making a requests over SSL. This host has a self signed certificate. If I reconfigure to point at our production servers with a proper ca certificate we do not get this exception and everything works as expected.

Note: we have added the self signed certificate to our JDK truststore.

Unfortunately it is a internal test host, so I cannot easily send you an example, but I will try to put together a self contained example.

```
[ERROR] [11/01/2013 13:23:02.999] [example-akka.actor.default-dispatcher-13] [akka://example/user/IO-HTTP/group-0/4] null
java.lang.IllegalStateException
    at spray.io.SslTlsSupport$$anon$1$$anon$2.spray$io$SslTlsSupport$$anon$$anon$$verify(SslTlsSupport.scala:361)
    at spray.io.SslTlsSupport$$anon$1$$anon$2$$anon$6$$anonfun$2.apply(SslTlsSupport.scala:92)
    at spray.io.SslTlsSupport$$anon$1$$anon$2$$anon$6$$anonfun$2.apply(SslTlsSupport.scala:81)
    at spray.io.SslTlsSupport$$anon$1$$anon$2.process(SslTlsSupport.scala:232)
    at spray.io.DynamicPipelines$$anonfun$eventPipeline$1.apply(Pipelines.scala:48)
    at spray.io.DynamicPipelines$$anonfun$eventPipeline$1.apply(Pipelines.scala:48)
    at spray.io.RawPipelineStage$$anon$3$$anonfun$2.apply(Pipelines.scala:115)
    at spray.io.RawPipelineStage$$anon$3$$anonfun$2.apply(Pipelines.scala:115)
    at spray.io.TickGenerator$$anon$1$$anon$2$$anonfun$1.apply(TickGenerator.scala:41)
    at spray.io.TickGenerator$$anon$1$$anon$2$$anonfun$1.apply(TickGenerator.scala:38)
    at spray.io.ConnectionHandler$$anonfun$running$1.applyOrElse(ConnectionHandler.scala:58)
    at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:165)
    at akka.actor.ActorCell.receiveMessage(ActorCell.scala:425)
    at akka.actor.ActorCell.invoke(ActorCell.scala:386)
    at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:230)
    at akka.dispatch.Mailbox.run(Mailbox.scala:212)
    at akka.dispatch.ForkJoinExecutorConfigurator$MailboxExecutionTask.exec(AbstractDispatcher.scala:506)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:262)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:975)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1478)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:104)
[ERROR] [11/01/2013 13:23:03.000] [example-akka.actor.default-dispatcher-7] [akka://example/user/IO-HTTP/group-0/4] Monitored actor [Actor[akka://example/system/IO-TCP/selectors/$a/8]] terminated (akka.actor.DeathPactException)
[WARN] [11/01/2013 13:23:03.005] [example-akka.actor.default-dispatcher-9] [akka://example/user/IO-HTTP/group-0/4] event pipeline: dropped Connected(example.com/172.29.130.33:443,/10.5.18.105:57335)
```

--- a/spray-io/src/main/scala/spray/io/SslTlsSupport.scala
+++ b/spray-io/src/main/scala/spray/io/SslTlsSupport.scala
@@ -83,13 +83,13 @@ object SslTlsSupport {
                 if (tracing) log.debug("Received {} inbound bytes in defaultState", data.size)
                 enqueueInboundBytes(data)
                 decrypt()
-                if (encryptedBytesPending) {
-                  sendEncryptedBytes()
+                if (encryptedBytesPending || isOutboundDone) {
+                  sendEncryptedBytes() // might send an empty Tcp.Write, but always triggers a pending ACK
                   become {
                     if (isOutboundDone) finishingClose(closedEvent)
                     else waitingForAck(remainingOutgoingData, closedEvent)
                   }
-                } else verify(!isOutboundDone)
+                }
               case Tcp.PeerClosed     ⇒ receivedUnexpectedPeerClosed()
               case x: Tcp.ErrorClosed ⇒ eventPL(x) // is there anything we need to close in this case?
               case x @ (_: Tcp.ConnectionClosed | WriteChunkAck) ⇒
@@ -208,10 +208,8 @@ object SslTlsSupport {
 
           def receivedUnexpectedPeerClosed(): Unit = {
             log.debug("Received unexpected Tcp.PeerClosed, invalidating SSL session")
-            try {
-              engine.closeInbound() // invalidates SSL session and should throw SSLException
-              throw new IllegalStateException("No SSLException after unexpected Tcp.PeerClosed")
-            } catch { case e: SSLException ⇒ } // ignore warning about truncation attack
+            try engine.closeInbound() // invalidates SSL session and should throw SSLException
+            catch { case e: SSLException ⇒ } // ignore warning about truncation attack
             become(finishingClose(Some(Tcp.ErrorClosed("Peer closed SSL connection prematurely")), Tcp.Close))
           }
 
@@ -341,7 +339,7 @@ object SslTlsSupport {
 
           def isOutboundDone: Boolean =
             if (engine.isInboundDone) {
-              // our pumping logic should make sure that we immediately outbound close and flush
+              // our pumping logic should make sure that we immediately outbound close
               // after having detected an inbound close
               verify(engine.isOutboundDone)
               true

